### PR TITLE
kubeadm: refactor discovery behind an interface

### PR DESCRIPTION
**Pull request by mikedanese**
This adds support for alternative discovery methods using discovery urls. It is a breaking change. This is a WIP.

Example usage:
```
$ kubeadm init --discovery token://
$ kubeadm join --discovery token://c05de9:ab224260fb3cd718@192.168.0.1:6555,191.168.0.2:6443
$ kubeadm join --discovery file:///etc/kubernetes/cluster.json
$ kubeadm join --discovery https://storage.google.apis.com/kube-discovery/98ea6e4/kubeconfig.json
```

@kubernetes/sig-cluster-lifecycle

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36462)

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: We're removing external etcd support?

**Comment by luxas** [comment]
@mikedanese verify-bolierplate failed, please fix the boilerplates
Also, the new discovery test needs an owner

File: build-tools/debs/kubeadm-10.conf
Comment (dgoodwin) [inline build-tools/debs/kubeadm-10.conf]: Is .deb build stuff moving into the main repo? (rpm as well?)

**Comment by luxas** [comment]
@mikedanese Why did you include the bazel things?
Since we're gonna split out kubeadm **_very soon_** (when master opens again), it seems unnecessary
Please remove unless there is a really good reason

File: cmd/kubeadm/app/cmd/init.go
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Oh, right.

File: cmd/kubeadm/app/util/tokens_test.go
Comment (mikedanese) [inline cmd/kubeadm/app/util/tokens_test.go]: WIP, will read

File: cmd/kubeadm/app/cmd/init.go
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/init.go]: WIP, will read

**Comment by mikedanese** [comment]
@luxas, I was using it for testing. Those commits won't be in the final PR

**Comment by luxas** [comment]
Ok, cool
Inform us when we should review :)

The overall concept LGTM

**Comment by luxas** [comment]
@mikedanese Ready for review :)?

**Comment by mikedanese** [comment]
@luxas Hopefully by tomorrow

**Comment by mikedanese** [comment]
@kubernetes/sig-cluster-lifecycle please give this a review. I'm going to run e2e tests on this tomorrow.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (dgoodwin) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: I do notice this leads to d.v.Token.Token etc, we could perhaps consider token redundant and go with Token.ID and Token.Secret.

File: cmd/kubeadm/app/discovery/token/token.go
Comment (dgoodwin) [inline cmd/kubeadm/app/discovery/token/token.go]: Neat idea.

Comment (dgoodwin) [inline cmd/kubeadm/app/discovery/token/token.go]: Neat idea.

File: cmd/kubeadm/app/master/tokens.go
Comment (dgoodwin) [inline cmd/kubeadm/app/master/tokens.go]: I will have some merge conflicts to deal with here on the token cli PR but we should probably merge this one first, and then I will update on my end.

File: cmd/kubeadm/app/preflight/checks.go
Comment (dgoodwin) [inline cmd/kubeadm/app/preflight/checks.go]: Compile error here:

```
# k8s.io/kubernetes/cmd/kubeadm/app/preflight
cmd/kubeadm/app/preflight/checks.go:248: cfg.Discovery.Token.Port undefined (type *kubeadm.TokenDiscovery has no field or method Port)
cmd/kubeadm/app/preflight/checks.go:278: cfg.APIPort undefined (type *kubeadm.NodeConfiguration has no field or method APIPort)
cmd/kubeadm/app/preflight/checks.go:279: cfg.Discovery.Token.Port undefined (type *kubeadm.TokenDiscovery has no field or method Port)
```

File: cmd/kubeadm/app/util/tokens.go
Comment (dgoodwin) [inline cmd/kubeadm/app/util/tokens.go]: Not having this method println: 👍

**Review by dgoodwin — CHANGES_REQUESTED** [review]
Ok made a pass through, I'm quite new to Go but everything looks ok aside from the problems noted. Would like to give it some manual testing but some compile issues preventing for now, will take another look as soon as they're ready.

File: cmd/kubeadm/app/discovery/flags_test.go
Comment (mikedanese) [inline cmd/kubeadm/app/discovery/flags_test.go]: Fixed.

**Comment by dgoodwin** [comment]
```
$ kubeadm init   
Running pre-flight checks
<master/tokens> accepted provided token
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4781b5]

goroutine 1 [running]:
panic(0x20f0480, 0xc42000e0c0)
        /home/dgoodwin/.gvm/gos/go1.7/src/runtime/panic.go:500 +0x1a1
k8s.io/kubernetes/cmd/kubeadm/app/util.BearerToken(0x0, 0x13, 0x1c0)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/util/tokens.go:77 +0x55
k8s.io/kubernetes/cmd/kubeadm/app/master.CreateTokenAuthFile(0x0, 0x1fa9380, 0xc42041c42d)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/master/tokens.go:56 +0x22a
k8s.io/kubernetes/cmd/kubeadm/app/cmd.(*Init).Run(0xc4204242f8, 0x35e2100, 0xc420026010, 0x0, 0xc4204242f8)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:208 +0x4a
k8s.io/kubernetes/cmd/kubeadm/app/cmd.NewCmdInit.func1(0xc4202ec6c0, 0x3669188, 0x0, 0x0)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go:79 +0xa6
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc4202ec6c0, 0x3669188, 0x0, 0x0, 0xc4202ec6c0, 0x3669188)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:603 +0x439
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4202ec240, 0xc4202ecd80, 0xc4202ec900, 0xc4202ec6c0)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:689 +0x367
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4202ec240, 0xc42000a7e0, 0x35e20c0)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:648 +0x2b
k8s.io/kubernetes/cmd/kubeadm/app.Run(0x0, 0x0)
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:39 +0x11f
main.main()
        /home/dgoodwin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:28 +0x22
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: This is hard to read, and it might produce the wrong outcome I think
If Token and File is specified it will be `false == false == true` => `false` which will pass the validation.

Please rewrite

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: `/Discovery/The discovery method kubeadm will use for connecting nodes to the master/` or similar

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Only if `i.cfg.Discovery.Token` is non-nil here I suppose, check for that

File: cmd/kubeadm/app/cmd/join.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: Same as above

File: cmd/kubeadm/app/cmd/join.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: This should be moved inside the if-statement I suppose?

File: cmd/kubeadm/app/cmd/join.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: What's this?

File: cmd/kubeadm/app/discovery/discovery.go
Comment (luxas) [inline cmd/kubeadm/app/discovery/discovery.go]: This does seem like WIP

File: cmd/kubeadm/app/util/tokens.go
Comment (luxas) [inline cmd/kubeadm/app/util/tokens.go]: I'm not saying you have to do it now, but can you move this file to a separate package for clarity (`cmd/kubeadm/app/util/token`) and import it as `tokenutil` where it's then used?

File: cmd/kubeadm/app/node/csr.go
Comment (luxas) [inline cmd/kubeadm/app/node/csr.go]: Can't we just squash these two methods together or is that yet to be done in a follow up?
If that's the case, add a TODO

**Review by luxas — COMMENTED** [review]
Looks good mostly, but still there are some places where it seems like you just forgot to implement things.

Ping me for a second review when fixed

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (mikedanese) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: `false == false == true => true` actually. This fails only for `true == true == true -> false` I think the correct expression is actually `(a == b == c) != (a && b && c)` but I'll just count the number of nils :)

File: cmd/kubeadm/app/node/csr.go
Comment (mikedanese) [inline cmd/kubeadm/app/node/csr.go]: TODO is to delete PerformTLSBootstrapDeprecated and move token bootstrap to PerformTLSBootstrap. I don't want to do that now but will add a todo.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: Be consistent here with all the above fields and do `json:"https"` here and below as well

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: But this isn't fixed still, right?

**Review by luxas — COMMENTED** [review]
I had one more comment now + plus nearly every comment from the previous review :)

Ping when fixed

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (mikedanese) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: No It's not fixed yet

**Comment by mikedanese** [comment]
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gke e2e test this

**Comment by mikedanese** [comment]
@k8s-bot gce etcd3 e2e test this

File: cmd/kubeadm/app/discovery/discovery.go
Comment (mikedanese) [inline cmd/kubeadm/app/discovery/discovery.go]: I'm defering implementing the discovery methods to another PR since this is already very big.

File: cmd/kubeadm/app/cmd/join.go
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/join.go]: It's a todo to move token discovery behind the discovery interface witch involved deleteing ConnectionInfo. I don't want to do that here.

File: cmd/kubeadm/app/util/tokens.go
Comment (mikedanese) [inline cmd/kubeadm/app/util/tokens.go]: all token stuff should be moved to cmd/kubeadm/app/discovery/token. Tokens have leaked and now permeate all of the kubeadm codebase which is very bad.

**Comment by mikedanese** [comment]
```
SSSS

Summarizing 1 Failure:

[Fail] ThirdParty resources [Flaky] [Disruptive] Simple Third Party [It] creating/deleting thirdparty objects works [Conformance] 
/usr/local/google/home/mikedanese/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/third-party.go:173

Ran 116 of 475 Specs in 3059.391 seconds
FAIL! -- 115 Passed | 1 Failed | 0 Pending | 359 Skipped --- FAIL: TestE2E (3059.41s)
FAIL
```

**Comment by mikedanese** [comment]
@luxas @dgoodwin PTAL now that the release is cut. This is passing e2e tests.

File: cmd/kubeadm/app/node/csr.go
Comment (luxas) [inline cmd/kubeadm/app/node/csr.go]: TODO and a nit: this should be `[node]` now

File: cmd/kubeadm/app/util/tokens.go
Comment (luxas) [inline cmd/kubeadm/app/util/tokens.go]: Also a future TODO

**Review by luxas — APPROVED** [review]
LGTM, only pointed out some places that are yet to be fixed in the future + those places I've pointedd out in earlier reviews

@dgoodwin PTAL if you want, otherwise I'm gonna lgtm this

**Comment by dgoodwin** [comment]
LGTM

**Pull request merged**
